src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,260 @@
+        {
+            _capacity = capacity > 0 ? capacity : throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException(nameof(capacity)));
+            _map = new ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>>(comparer ?? EqualityComparer<TKey>.Default);
+
+            var thread = new Thread(new ThreadStart(OnStart));
+            thread.IsBackground = true;
+            thread.Start();
+
+            Task timerTask = RemoveExpiredValuesPeriodically(TimeSpan.FromMinutes(5));
+        }
+
+        private void OnStart()
+        {
+            while (true)
+            {
+                _eventQueue.Take().Invoke();
+            }
+        }
+
+        public bool Contains(TKey key)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            return _map.ContainsKey(key);
+        }
+
+        public int RemoveExpiredValues()
+        {
+            int numItemsRemoved = 0;
+            var node = _doubleLinkedList.First;
+            while (node != null)
+            {
+                var nextNode = node.Next;
+                if (node.Value.ExpirationTime < DateTime.UtcNow)
+                {
+                    _doubleLinkedList.Remove(node);
+                    _map.TryRemove(node.Value.Key, out _);
+                    numItemsRemoved++;
+                }
+
+                node = nextNode;
+            }
+
+            return numItemsRemoved;
+        }
+
+        async Task RemoveExpiredValuesPeriodically(TimeSpan interval)
+        {
+            while (true)
+            {
+                _eventQueue.Add(() => RemoveExpiredValues());
+                await Task.Delay(interval).ConfigureAwait(false);
+            }
+        }
+
+        public void SetValue(TKey key, TValue value)
+        {
+            SetValue(key, value, DateTime.MaxValue);
+        }
+
+        public bool SetValue(TKey key, TValue value, DateTime? expirationTime)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (value == null)
+                throw LogHelper.LogArgumentNullException(nameof(value));
+
+            if (expirationTime == null)
+                throw LogHelper.LogArgumentNullException(nameof(expirationTime));
+
+            // item already expired
+            if (expirationTime < DateTime.UtcNow)
+                return false;
+
+            // just need to update value and move it to the top
+            if (_map.ContainsKey(key))
+            {
+                // make sure node hasn't been removed by a different thread
+                if (_map.TryGetValue(key, out var cacheItem))
+                    _eventQueue.Add(() => _doubleLinkedList.Remove(cacheItem));
+
+                // add a new item regardless of whether the old item was removed or not
+                _eventQueue.Add(() => _doubleLinkedList.AddFirst(new LRUCacheItem<TKey, TValue>(key, value, expirationTime)));
```

Comment:
Should we batch Remove and AddFirst in set and get to keep these operations as one and not having to worry about what impact an another thread adding another event in between these two would cause?

Reply:
This makes sense to me! Fixed. :) 

I didn't group them initially since I figured the exact order didn't really matter (e.g. we're okay as long as we're close enough to LRU), but I do think it's better to be more deterministic about this if we can.

Commit: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1575/commits/45c168bd46aa16f03a5e5f83c8b9b2b8cef9b070